### PR TITLE
Fixes #37398 - Extend TableIndexPage and TableHooks with new capabilities for All Hosts page

### PR DESCRIPTION
**Pull request by jeremylenz**
This adds several features to TableIndexPage and TableHooks, enabling https://github.com/Katello/katello/pull/10962 to work. Ideally the two PRs should be reviewed together.

- add the option to use `idColumnName` to `<RowSelectTd>`. This is so you can optionally pass in an alternate `idColumn` to `useBulkSelect`.
- Add the option to supply your own pagination component in `<Table>`. When using the `<TableIndexPage>` without specifying children, it was causing some issues with the bottom pagination not being in sync or making API requests. Allowing the parent to specify the bottomPagination solves this, because you can now guarantee that the top and bottom pagination have the same data.
- Add a `hasInteracted` state to `useBulkSelect`. This is so you can delay showing validations (e.g. for forms and wizards ;) until the user has actually interacted with the component. It's bad UX to throw up a validation message such as "You must select at least one X!!" before the user has even gotten a chance to do anything. I had tried adding this to the pagination functions in Katello, but this solution is much more elegant. 
- Add a `pushToHistory` option to `useSetParamsAndAPIAndSearch`. When you are using an embedded table (e.g. one that should not keep itself in sync with URL params for search and pagination), you need to be able to skip changing the browser URL.
- Add `updateParamsByUrl` to `<TableIndexPage>` for the same reason.
- Add `restrictedSearchQuery` to `<TableIndexPage>`. This allows you to restrict all table rows to ones matching the search you specify here. This restricted search query will not show up visually in the search input, but will be added to the user search in the API request, causing every search to be "within" the restricted search query's results. Usage example: I select 50 hosts and bring up the package install wizard. The package install wizard has the 50 hosts I selected. I am able to search within those and deselect if desired, but I cannot add any hosts to the selection.

### CODE REVIEW

**Comment by jeremylenz** [comment]
Updated package.json with `"@theforeman/vendor": "^13.1.0"` to enable the use of `@patternfly/react-core/next`.

**Comment by evgeni** [comment]
Packit will still fail, as 13.1.0 ain't packaged yet, but we'll get there

**Comment by ianballou** [comment]
Testing this PR via https://github.com/Katello/katello/pull/10962 showed no technical issues.

**Comment by evgeni** [comment]
@jeremylenz you dropped the use of `@patternfly/react-core/next` again?

**Comment by jeremylenz** [comment]
@evgeni Nope, that use is in the Katello PR.

**Review by parthaa — APPROVED** [review]
Happy with the changes here. Tested it with Katello PR. Works well :+1:

**Comment by jeremylenz** [comment]
Re: React on plugins katello/katello failures, see https://github.com/Katello/katello/pull/10962#discussion_r1603201718

**Comment by jeremylenz** [comment]
I plan to wait til after stabilization week to merge this. Will need an ack from @theforeman/packaging then.

**Comment by jeremylenz** [comment]
Foreman is now branched. @theforeman/packaging are we good here?

**Review by evgeni — DISMISSED** [review]

**Comment by evgeni** [comment]
(that ack is purely packaging, I did not look at the code)

**Comment by MariaAga** [comment]
@jeremylenz I'm starting to look over the PR and it took me a while to find the actions buttons since there are 4 dropdowns in the page, not sure if its for this pr, but should we combine some dropdowns? 
Screenshot from 2024-06-03 20-34-55(1)

**Comment by jeremylenz** [comment]
> but should we combine some dropdowns?

All those are in the mockups :shrug: 

They do each have their own purpose:
1) general actions you can do without selecting hosts
2) REX jobs (self-explanatory and not a kebab anyway)
3) actions you can do with multiple hosts selected
4) actions you can do only for the single host in the table row

Given this, I'm not sure it makes sense to combine any of them IMHO.

**Comment by ekohl** [comment]
> REX jobs (self-explanatory and not a kebab anyway)

Why is REX a special case? Aren't REX jobs just a regular bulk action you can use when you select hosts?

**Comment by jeremylenz** [comment]
> Why is REX a special case? Aren't REX jobs just a regular bulk action you can use when you select hosts?

"Schedule a Job" is a split button with a dropdown. You can click the button itself without opening the dropdown, which takes you directly to a blank job wizard with the selected hosts pre-filled, or you can select an item from the dropdown, which runs a specific job on those hosts. AFAIK this is similar to the 'Schedule a Job' on the legacy page, so to me it would be inconsistent to remove it and move all the actions somewhere else. It also conforms to the designs we have, which we probably shouldn't alter too much without getting more UX input.

**Comment by ekohl** [comment]
> "Schedule a Job" is a split button with a dropdown. You can click the button itself without opening the dropdown, which takes you directly to a blank job wizard with the selected hosts pre-filled, or you can select an item from the dropdown, which runs a specific job on those hosts. AFAIK this is similar to the 'Schedule a Job' on the legacy page, so to me it would be inconsistent to remove it and move all the actions somewhere else. It also conforms to the designs we have, which we probably shouldn't alter too much without getting more UX input.

In the existing UI scheduling a job is just a regular bulk host action. It only works if you have selected at least one host.

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: since this is repeating 3 times in this file can it be in const in case we ever change it?

**Review by MariaAga — CHANGES_REQUESTED** [review]

**Comment by MariaAga** [comment]
Also the restrictedSearchQuery doesn't work well, I think the issue is in this pr and not the katello one, but when I search for `name ~ *` it adds more items to the list (in the package wizard host step)

**Comment by jeremylenz** [comment]
> Also the restrictedSearchQuery doesn't work well, I think the issue is in this pr and not the katello one, but when I search for `name ~ *` it adds more items to the list (in the package wizard host step)

Fixed in Katello. I had somehow stopped passing in `initialSelectedHosts` to `HostReview`.

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: Updated :+1:

**Comment by jeremylenz** [comment]
rebased

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: Meant also the page || 1.
Also actually why is this needed? the default value for page is 1 and perPage uses the settings perPage if not defined

**Comment by MariaAga** [comment]
> Fixed in Katello. I had somehow stopped passing in initialSelectedHosts to HostReview.

I updated both prs and can still select more hosts than the initial selection with `name ~ *` if its by design then please change the wording of `Review and optionally exclude hosts from your selection.`  since hosts can be excluded and also *added*

**Comment by jeremylenz** [comment]
Sorry, I had committed locally but forgot to push in the Katello PR :flushed:  try again now?

**Comment by jeremylenz** [comment]
Removed the `Number()` uses. (could have sworn I needed them but it seems to work okay without them?) Also fixed a bug where the selectedCount would be incorrect in select all mode if there are no search results.

**Review by MariaAga — APPROVED** [review]
Looks good to me, but I would like to see that it doesn't break the remidiation wizard with the initialQuery change, and dont have openScap working yet. 
@ofedoren (or @jeremylenz if you have the plugin) Could you just open the openscap table to see that its fine?

**Comment by jeremylenz** [comment]
Looks like this also will need another @theforeman/packaging ack for some reason.

**Comment by ofedoren** [comment]
> I would like to see that it doesn't break the remidiation wizard with the initialQuery change

I don't think we use that much in the wizard, we also disabled searchbox there.

> Could you just open the openscap table to see that its fine?

I've checked the wizard and everything seems working just fine without any changes.

**Review by evgeni — APPROVED** [review]
:package:

**Pull request merged**
